extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -0,0 +1,706 @@
+    // buildCallback promise chain.
+    // If another ad is currently loading we only load ads that are currently
+    // in viewport.
+    const allowRender = allowRenderOutsideViewport(this.element, this.getWin());
+    if (allowRender !== true) {
+      return allowRender;
+    }
+    // Otherwise the ad is good to go.
+    return super.renderOutsideViewport();
+  }
+
+  /**
+   * To be overridden by network specific implementation indicating if element
+   * (and environment generally) are valid for sending XHR rqueries.
+   * @return {boolean} where element is valid and ad request should be sent.  If
+   *    false, no ad request is sent and slot will be collapsed if possible.
+   */
+  isValidElement() {
+    return true;
+  }
+
+  /**
+   * Prefetches and preconnects URLs related to the ad using adPreconnect
+   * registration which assumes ad request domain used for 3p is applicable.
+   * @override
+   */
+  preconnectCallback(unusedOnLayout) {
+    const preconnect = adPreconnect[this.element.getAttribute('type')];
+    // NOTE(keithwrightbos): using onLayout to indicate if preconnect should be
+    // given preferential treatment.  Currently this would be false when
+    // relevant (i.e. want to preconnect on or before onLayoutMeasure) which
+    // causes preconnect to delay for 1 sec (see custom-element#preconnect)
+    // therefore hard coding to true.
+    // NOTE(keithwrightbos): Does not take isValidElement into account so could
+    // preconnect unnecessarily, however it is assumed that isValidElement
+    // matches amp-ad loader predicate such that A4A impl does not load.
+    if (typeof preconnect == 'string') {
+      this.preconnect.url(preconnect, true);
+    } else if (preconnect) {
+      preconnect.forEach(p => {
+        this.preconnect.url(p, true);
+      });
+    }
+  }
+
+  /** @override */
+  onLayoutMeasure() {
+    if (this.intersectionObserver_) {
+      this.intersectionObserver_.fire();
+    }
+    if (this.layoutMeasureExecuted_ || !verifySignatureIsAvailable()) {
+      // onLayoutMeasure gets called multiple times.
+      return;
+    }
+    this.layoutMeasureExecuted_ = true;
+    if (isPositionFixed(this.element, this.getWin())) {
+      user.error('<amp-ad> is not allowed to be placed in elements with ' +
+                 'position:fixed: %s', this.element);
+      return;
+    }
+    // OnLayoutMeasure can be called when page is in prerender so delay until
+    // visible.  Assume that it is ok to call isValidElement as it should
+    // only being looking at window, immutable properties (i.e. location) and
+    // its element ancestry.
+    if (!this.isValidElement()) {
+      // TODO(kjwright): collapse?
+      user.warn('Amp Ad', 'Amp ad element ignored as invalid', this.element);
+      return;
+    }
+
+    // Increment unique promise ID so that if its value changes within the
+    // promise chain due to cancel from unlayout, the promise will be rejected.
+    this.promiseId_++;
```

Comment:
`onLayoutMeasure` is also called every time the viewport resizes.

- It might be best to increment this only in `#unlayoutCallback`. It could also be a boolean, with it being set to `true` in `#layoutCallback` and `false` in `#unlayoutCallback`.
  - Which would also prevent very large promise chains that are essentially useless.